src/encoder.rs

```diff
@@ -175,9 +175,14 @@ impl<W: Write> Encoder<W> {
     /// sample values based on the previous. For a potentially better compression ratio, at the
     /// cost of more complex processing, try out [`FilterType::Paeth`].
     ///
+    /// A filter type of `None` sets the filter to Adaptive. Adaptive filtering
+    /// attempts to select the best filter for each line based on a heuristic
+    /// described in the PNG standard. This can result in smaller file sizes
+    /// but is not guaranteed to produce the best compression ratio.
+    ///
     /// [`FilterType::Sub`]: enum.FilterType.html#variant.Sub
     /// [`FilterType::Paeth`]: enum.FilterType.html#variant.Paeth
-    pub fn set_filter(&mut self, filter: FilterType) {
+    pub fn set_filter(&mut self, filter: Option<FilterType>) {
```

Comment:
I added an `AdaptiveFilterType` enum which contains Adaptive and NonAdaptive. I do agree it seems clearer than the Option machinery especially as a user. I think this will allow more opportunity to customize in the future.

I added an `adaptive_filter` field to Encoder, Writer, and StreamWriter. StreamWriter wraps around a Writer but I'm not sure I can access the filter and adaptive_filter from the Writer. So I have to duplicate the fields in StreamWriter, the filter_type was already present.